### PR TITLE
phpstan: fixes for Mage_Catalog_Model_Product_Type_Abstract

### PR DESCRIPTION
**Pull request by sreichel**
### Description (*)

Fixes call to undefined method on Mage_Catalog_Model_Product_Type_Abstract ...

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit.php [hunk 1]
```diff
@@ -234,15 +234,18 @@ public function getAttributeSetName()
...
+            /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
+            $productType = $product->getTypeInstance(true);
+            $superAttributes = $productType->getUsedProductAttributeIds($this->getProduct());
+            if (!$superAttributes) {
+                $superAttributes = false;
+            }
         }
 
-        return !$this->getProduct()->isConfigurable() || $superAttributes !== false;
+        return !$product->isConfigurable() || $superAttributes !== false;
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit.php:248]: Not sure if this is easier to read:

```suggestion
        $product = $this->getProduct();
        if ($product->isConfigurable()) {
            /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
            $productType = $product->getTypeInstance(true);
            return !empty($productType->getUsedProductAttributeIds($product));
        } else {
            return true;
        }
```

If it is not, line 242 `$superAttributes = $productType->getUsedProductAttributeIds($this->getProduct());` can be shortened to `$superAttributes = $productType->getUsedProductAttributeIds($product);`

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php [hunk 2]
```diff
@@ -58,45 +58,47 @@ protected function _prepareForm()
             ->setAttributeSetId($this->_getProduct()->getAttributeSetId())
             ->getAttributes();
 
+        /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
+        $productType = $this->_getProduct()->getTypeInstance(true);
+        $usedAttributes = $productType->getUsedProductAttributes($this->_getProduct());
+
         /* Standart attributes */
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:65]:
```suggestion
        /* Standard attributes */
        /** @var Mage_Eav_Model_Attribute $attribute */
```

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php [hunk 3]
```diff
@@ -108,7 +110,9 @@ protected function _prepareForm()
         }
 
         /* Configurable attributes */
-        $usedAttributes = $this->_getProduct()->getTypeInstance(true)->getUsedProductAttributes($this->_getProduct());
+        /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
+        $productType = $this->_getProduct()->getTypeInstance(true);
+        $usedAttributes = $productType->getUsedProductAttributes($this->_getProduct());
         foreach ($usedAttributes as $attribute) {
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:116]:
```suggestion
        /** @var Mage_Eav_Model_Attribute $attribute */
        foreach ($usedAttributes as $attribute) {
```

File: app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php [hunk 4]
```diff
@@ -98,11 +98,12 @@ protected function _initProduct()
...
 
         $attributes = $this->getRequest()->getParam('attributes');
-        if ($attributes && $product->isConfigurable() &&
-            (!$productId || !$product->getTypeInstance()->getUsedProductAttributeIds())) {
-            $product->getTypeInstance()->setUsedProductAttributeIds(
-                explode(",", base64_decode(urldecode($attributes)))
-            );
+        if ($attributes && $product->isConfigurable()) {
+            if (!$productId || !$product->getTypeInstance()->getUsedProductAttributeIds()) {
+                $product->getTypeInstance()->setUsedProductAttributeIds(
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php:103]:
```suggestion
            /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
            $productType = $product->getTypeInstance();
            if (!$productId || !$productType->getUsedProductAttributeIds()) {
                $productType->setUsedProductAttributeIds(
```

File: app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php [hunk 5]
```diff
@@ -207,10 +208,12 @@ public function newAction()
             $this->loadLayout('popup');
         } else {
             $_additionalLayoutPart = '';
-            if ($product->getTypeId() == Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE
-                && !($product->getTypeInstance()->getUsedProductAttributeIds()))
-            {
-                $_additionalLayoutPart = '_new';
+            if ($product->getTypeId() == Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE) {
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php:211]:
```suggestion
            if ($product->getTypeId() === Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE) {
```

File: app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php [hunk 6]
```diff
@@ -249,10 +252,12 @@ public function editAction()
         Mage::dispatchEvent('catalog_product_edit_action', ['product' => $product]);
 
         $_additionalLayoutPart = '';
-        if ($product->getTypeId() == Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE
-            && !($product->getTypeInstance()->getUsedProductAttributeIds()))
-        {
-            $_additionalLayoutPart = '_new';
+        if ($product->getTypeId() == Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE) {
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/controllers/Catalog/ProductController.php:255]:
```suggestion
        if ($product->getTypeId() === Mage_Catalog_Model_Product_Type::TYPE_CONFIGURABLE) {
```

**Review by kiatng — COMMENTED** [review]
Reviewed Mage/Adminhtml. To do the rest.

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php (hunk 3 above)
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:117]: Why this? `Mage_Catalog_Model_Product_Type_Configurable::getUsedProductAttributes()` returns ...

```
     * @return Mage_Catalog_Model_Product_Type_Configurable_Attribute[]
```

pr-2605

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:117]: My bad. You are right, it should be `Mage_Catalog_Model_Product_Type_Configurable_Attribute`.

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:117]: Wait, I think `$attribute` is  `Mage_Catalog_Model_Resource_Eav_Attribute`:

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:117]: Yep. Is it possibie to update that suggestion? Return turn should be changed here:

```
     * @param  Mage_Catalog_Model_Product $product
     * @return Mage_Catalog_Model_Resource_Eav_Attribute[]
     */
    public function getUsedProductAttributes($product = null)
```

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php [hunk 7]
```diff
@@ -108,7 +111,9 @@ protected function _prepareForm()
...
```
[the remark this answers is not in the document]
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:116]: With last commit this comments could be removed?

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php [hunk 8]
```diff
@@ -58,45 +58,48 @@ protected function _prepareForm()
             ->setAttributeSetId($this->_getProduct()->getAttributeSetId())
             ->getAttributes();
 
-        /* Standart attributes */
+        /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
+        $productType = $this->_getProduct()->getTypeInstance(true);
+        $usedAttributes = $productType->getUsedProductAttributes($this->_getProduct());
+
+        /* Standard attributes */
+        /** @var Mage_Eav_Model_Attribute $attribute */
```
[the remark this answers is not in the document]
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php]: With last commit (https://github.com/OpenMage/magento-lts/commit/bbf361b681caa1e72d5f6a3dfbdea070ffea16e6) this comments could be removed?

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit.php (hunk 1 above)
Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit.php]: Used the shortend version (for now).

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit.php]: @kiatng i also thought about make it simplier ... eg first ...

```
if (!$product->isConfigurable()) {
    return true;
}
...
```

... but i thought its better to keep changes as simple as possible, I did not expect your careful reviews :+1: :100: 

There is a lot of code that could be made simpler. If you see something like this ... maybe create a PR?

File: app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php [hunk 9]
```diff
@@ -108,7 +110,9 @@ protected function _prepareForm()
         }
 
         /* Configurable attributes */
-        $usedAttributes = $this->_getProduct()->getTypeInstance(true)->getUsedProductAttributes($this->_getProduct());
+        /** @var Mage_Catalog_Model_Product_Type_Configurable $productType */
+        $productType = $this->_getProduct()->getTypeInstance(true);
+        $usedAttributes = $productType->getUsedProductAttributes($this->_getProduct());
```
Comment (kiatng) [inline app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:115]: Just asking, is re-declaring `$productType` and `$usedAttributes` for BC? I do not see how the values of these two vars can changed from lines 62-63.

Comment (sreichel) [reply app/code/core/Mage/Adminhtml/Block/Catalog/Product/Edit/Tab/Super/Config/Simple.php:115]: I just fixed the lines mentioned at phpstan baseline. I did not check for duplicate code. :+1:

**Review by kiatng — APPROVED** [review]

**Review by fballiano — APPROVED** [review]

**Comment by shaughnlgh** [comment]
A bug has been introduced with this commit: https://github.com/OpenMage/magento-lts/commit/37642edbac1c2695269c1653c42aaed72f36a136#r85358437

**Pull request merged**